File: tests/PREVAC_TESTS.cc
#include <vector>
#include <boost/asio.hpp>
#include <iostream>
#include <fstream>
#include <cstring>
#include <unistd.h>
#include <fcntl.h>
#include <termios.h>

namespace asio = boost::asio;
using namespace std;

asio::io_service io;
asio::serial_port serial(io);

int RS485;

vector<unsigned char> calculate_crc(const vector<unsigned char>& data) {
    unsigned short CRC = 0xFFFF;
    for (unsigned char byte : data) {
        CRC ^= byte;
        for (int i = 0; i < 8; ++i) {
            bool carry = CRC & 1;
            CRC >>= 1;
            if (carry) {
                CRC ^= 0xA001;
            }
        }
    }
    vector<unsigned char> result;
    result.push_back((CRC >> 8) & 0xFF);
    result.push_back(CRC & 0xFF);
    return result;
}

tuple<int, int, int, vector<unsigned char>> read_response() {
    vector<unsigned char> A(1), B(1), C_data(1), CRC_received(2);
    read(RS485, &A, 1);
    read(RS485, &B, 1);
    read(RS485, &C_data, 1);
    std::cerr << &A << &B << &C_data << std::endl;
    int address = static_cast<int>(A[0]);
    int function_code = static_cast<int>(B[0]);
    int C = static_cast<int>(C_data[0]);
    vector<unsigned char> data;
    data.resize(C);
    read(RS485, data.data(), C);
    read(RS485, CRC_received.data(), 4);
    return make_tuple(address, function_code, C, data);
}

int main() {
    RS485 = open("/dev/serial/by-id/usb-1a86_USB_Serial-if00-port0", O_RDWR | O_NOCTTY | O_NDELAY);
    if (RS485 == -1) {
        std::cerr << "Error al abrir el puerto serial." << std::endl;
        return 1;
    }

    struct termios options;
    tcgetattr(RS485, &options);
    cfsetispeed(&options, B9600);
    cfsetospeed(&options, B9600);
    options.c_cflag |= (CLOCAL | CREAD);
    options.c_cflag &= ~PARENB;
    options.c_cflag &= ~CSTOPB;
    options.c_cflag &= ~CSIZE;
    options.c_cflag |= CS8;
    options.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG);
    options.c_oflag &= ~OPOST;
    options.c_cc[VMIN] = 0;
    options.c_cc[VTIME] = 10;
    tcsetattr(RS485, TCSANOW, &options);

    // Obtener la entrada del usuario
    int strt_mem, win_amount;
    cout << "First word address: ";
    cin >> strt_mem;
    cout << "Amount of windows to read: ";
    cin >> win_amount;

    // Calcular los valores a enviar
    int chr3 = (strt_mem >> 8) & 0xFF;
    int chr4 = strt_mem & 0xFF;
    int chr5 = (win_amount >> 8) & 0xFF;
    int chr6 = win_amount & 0xFF;

    vector<unsigned char> data = {1, 3, static_cast<unsigned char>(chr3), static_cast<unsigned char>(chr4),
                                   static_cast<unsigned char>(chr5), static_cast<unsigned char>(chr6)};
    auto crc = calculate_crc(data);

    // Enviar datos al puerto serial
    write(RS485, data.data(), data.size());
    write(RS485, &crc, sizeof(crc));

    // Leer la respuesta del puerto serial
    auto response = read_response();

    // Imprimir la respuesta
    if (get<3>(response).size() > 0) {
        cout << "Address: " << get<0>(response) << endl;
        cout << "Function Code: " << get<1>(response) << endl;
        cout << "Data Length: " << get<2>(response) << endl;
        cout << "Data: ";
        for (unsigned char byte : get<3>(response)) {
            cout << static_cast<int>(byte) << " ";
        }
        cout << endl;
    } else {
        cout << "Error al leer la respuesta." << endl;
    }

    // Cerrar el puerto serial
    close(RS485);

    return 0;
}

File: tests/PREVAC_TESTS.py
import serial

RS232 = serial.Serial(port='/dev/serial/by-id/usb-1a86_USB2.0-Ser_-if00-port0', baudrate=9600)

def calculate_crc(data):
    CRC = 0xFFFF
    for byte in data:
        CRC ^= byte
        for _ in range(8):
            carry = CRC & 1
            CRC >>= 1
            if carry:
                CRC ^= 0xA001
    return CRC.to_bytes(2, 'big')

def read_response():
    A = RS232.read()
    B = RS232.read()
    address = int.from_bytes(A, 'big')
    function_code = int.from_bytes(B, 'big')
    C = int.from_bytes(RS232.read(), 'big')
    data = RS232.read(C)
    CRC_received = RS232.read(2)
    return address, function_code, C, data

strt_mem = int(input("First word address: "))  

chr3 = (strt_mem >> 8) & 0xFF
chr4 = strt_mem & 0xFF

chr0 = 0
chr1 = 1 # address
chr2 = 3 # read

win_amount = int(input("Amount of windows to read: ")) 

chr5 = (win_amount >> 8) & 0xFF
chr6 = win_amount & 0xFF

data = [chr1, chr2, chr3, chr4, chr5, chr6]
crc1, crc2 = calculate_crc(data)

Chr0 = chr0.to_bytes(3, 'big')
Chr1 = chr1.to_bytes(1, 'big')
Chr2 = chr2.to_bytes(1, 'big')
Chr3 = chr3.to_bytes(1, 'big')
Chr4 = chr4.to_bytes(1, 'big')
Chr5 = chr5.to_bytes(1, 'big')
Chr6 = chr6.to_bytes(1, 'big')
CRC1 = crc1.to_bytes(1, 'big')
CRC2 = crc2.to_bytes(1, 'big')

print(Chr3, Chr4, Chr5, Chr6) 

RS232.write(Chr1)
RS232.write(Chr2)
RS232.write(Chr3)
RS232.write(Chr4)
RS232.write(Chr5)
RS232.write(Chr6)
RS232.write(CRC2)
RS232.write(CRC1)

response = read_response()

if response:
    address, function_code, C, data = response
    print(f"Address: {address}")
    print(f"Function Code: {function_code}")
    print(f"Data Length: {C}")
    print(f"Data: {data}")
else:
    print("Error al leer la respuesta.")

RS232.close()

